src/hotspot/share/opto/memnode.cpp

```diff
@@ -3519,17 +3519,130 @@ Node *StoreNode::Ideal_masked_input(PhaseGVN *phase, uint mask) {
+// and (x << 31) << 2 is simplified into 0. This means that in any case, after the
+// left shift, we always have at least one bit of the original v.
+//
+// ### Case 1 : conIL == conIR
+// If we do the shift left then right by 24 bits, we get:
+// after << 24
+// +---------+------------------------+
+// | v[0..7] |           0            |
+// +---------+------------------------+
+//  31     24 23                      0
+// after >> 24
+// +------------------------+---------+
+// |        sign bit        | v[0..7] |
+// +------------------------+---------+
+//  31                     8 7        0
+// The non-rejected bits are the same before and after, so, indeed, simplifying is fine.
+//
+// ### Case 2: conIL > conIR == num_rejected_bits
+// We take conIL == 26 for this example.
+// after << 26
+// +---------+------------------------+
+// | v[0..5] |           0            |
+// +---------+------------------------+
+//  31     26 25                      0
+// after >> 24
+// +------------------+---------+-----+
+// |     sign bit     | v[0..5] |  0  |
+// +------------------+---------+-----+
+//  31               8 7       2 1   0
+// The non-rejected bits are the 8 lower one of (v << conIL - conIR).
+// The bits 6 and 7 of v have been thrown away after the shift left.
+// The simplification is still fine.
+//
+// ### Case 3: conIL > conIR < num_rejected_bits.
+// Let's say conIL == 26 and conIR == 22.
+// after << 26
+// +---------+------------------------+
+// | v[0..5] |           0            |
+// +---------+------------------------+
+//  31     26 25                      0
+// after >> 22
+// +------------------+---------+-----+
+// |     sign bit     | v[0..5] |  0  |
+// +------------------+---------+-----+
+//  31              10 9       4 3   0
+// The non-rejected bits are the 8 lower one of (v << conIL - conIR).
+// The bits 6 and 7 of v have been thrown away after the shift left.
+// The bits 4 and 5 of v are still present, but outside of the kept bits (the 8 lower ones).
+// The simplification is still fine.
+//
+// ### But! Case 4: conIL > conIR > num_rejected_bits.
+// Let's see that this case is not as easy to simplify.
+// Let's say conIL == 28 and conIR == 26.
+// after << 28
+// +---------+------------------------+
+// | v[0..3] |           0            |
+// +---------+------------------------+
+//  31     28 27                      0
+// after >> 26
+// +------------------+---------+-----+
+// |     sign bit     | v[0..3] |  0  |
+// +------------------+---------+-----+
+//  31               6 5       2 1   0
+// The non-rejected bits are made of
+// - 0-1 => 0
+// - 2-5 => the bits 0 to 3 of v
+// - 6-7 => the sign bit of v[0..3] (that is v[3])
+// Simplifying this as (X << 2) is not correct.
+// The condition conIR <= num_rejected_bits is indeed necessary.
+//
+// ### Conclusion:
+// Valid if conIL >= conIR <= num_rejected_bits
+//
+// We do not treat the case conIR > conIL here since the point of this function is
```

Comment:
```suggestion
// We do not treat the case conIL < conIR here since the point of this function is
```
Nit: just to keep the two values on the sides we have used up to now.

Reply:
Done.